File: src/runtime/writeIntegration.ts
import { spawnSync } from "node:child_process";

export function applyPatch(repoRoot: string, patch: string, allowedGlobs: string[]): void {
  applyPatches(repoRoot, [patch], allowedGlobs);
}

export function applyPatches(repoRoot: string, patches: string[], allowedGlobs: string[]): void {
  if (patches.length === 0) return;
  const combined = patches.join("\n");
  const changedPaths = patchChangedPaths(repoRoot, combined);
  validateWriteAllowlist(changedPaths, allowedGlobs);
  const check = spawnSync("git", ["apply", "--check", "--3way", "-"], { cwd: repoRoot, input: combined, encoding: "utf8", maxBuffer: 50 * 1024 * 1024 });
  if (check.status !== 0) throw new Error(check.stderr || check.stdout);
  const apply = spawnSync("git", ["apply", "--3way", "-"], { cwd: repoRoot, input: combined, encoding: "utf8", maxBuffer: 50 * 1024 * 1024 });
  if (apply.status !== 0) throw new Error(apply.stderr || apply.stdout);
}

export function patchChangedPaths(repoRoot: string, patch: string): string[] {
  const result = spawnSync("git", ["apply", "--numstat", "-"], { cwd: repoRoot, input: patch, encoding: "utf8", maxBuffer: 50 * 1024 * 1024 });
  if (result.status !== 0) throw new Error(result.stderr || result.stdout);
  return result.stdout
    .split(/\r?\n/)
    .filter(Boolean)
    .map((line) => line.split("\t").at(-1) ?? "")
    .filter(Boolean);
}

function validateWriteAllowlist(changedPaths: string[], allowedGlobs: string[]): void {
  for (const changedPath of changedPaths) {
    if (!allowedGlobs.some((glob) => pathMatches(changedPath, glob))) {
      throw new Error(`patch changes path outside manifest.permissions.write: ${changedPath}`);
    }
  }
}

function pathMatches(file: string, glob: string): boolean {
  if (glob === "**") return true;
  if (glob.endsWith("/**")) return file.startsWith(glob.slice(0, -2));
  if (!glob.includes("*")) return file === glob;
  const escaped = glob
    .replace(/[.+^${}()|[\]\\]/g, "\\$&")
    .replace(/\*\*/g, ".*")
    .replace(/\*/g, "[^/]*");
  return new RegExp(`^${escaped}$`).test(file);
}

File: src/write_integration.rs
use crate::snapshot::matches_glob;
use anyhow::{Result, bail};
use std::path::Path;
use std::process::{Command, Stdio};

pub fn apply_patches(repo_root: &Path, patches: &[String], allowed_globs: &[String]) -> Result<()> {
    if patches.is_empty() {
        return Ok(());
    }
    let combined = patches.join("\n");
    let changed_paths = patch_changed_paths(repo_root, &combined)?;
    for path in &changed_paths {
        if !allowed_globs.iter().any(|glob| matches_glob(path, glob)) {
            bail!("patch changes path outside manifest.permissions.write: {path}");
        }
    }
    run_git_apply(repo_root, &["apply", "--check", "--3way", "-"], &combined)?;
    run_git_apply(repo_root, &["apply", "--3way", "-"], &combined)?;
    Ok(())
}

pub fn patch_changed_paths(repo_root: &Path, patch: &str) -> Result<Vec<String>> {
    let output = run_git_apply(repo_root, &["apply", "--numstat", "-"], patch)?;
    Ok(String::from_utf8_lossy(&output)
        .lines()
        .filter_map(|line| line.split('\t').next_back())
        .filter(|item| !item.is_empty())
        .map(ToString::to_string)
        .collect())
}

fn run_git_apply(repo_root: &Path, args: &[&str], input: &str) -> Result<Vec<u8>> {
    let mut child = Command::new("git")
        .args(args)
        .current_dir(repo_root)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;
    {
        use std::io::Write;
        child
            .stdin
            .as_mut()
            .expect("stdin is piped")
            .write_all(input.as_bytes())?;
    }
    let output = child.wait_with_output()?;
    if !output.status.success() {
        bail!(
            "{}{}",
            String::from_utf8_lossy(&output.stderr),
            String::from_utf8_lossy(&output.stdout)
        );
    }
    Ok(output.stdout)
}
